src/extlibs/hrvo/hrvo_agent.cpp

```diff
@@ -234,222 +241,226 @@ void HRVOAgent::computeNewVelocity()
                 {
                     candidate.position_ =
-                        velocityObstacles_[i].apex_ + s * velocityObstacles_[i].side2_;
-
-                    if (candidate.position_.lengthSquared() < max_speed_ * max_speed_)
-                    {
-                        candidates_.insert(std::make_pair(
-                            (pref_velocity_ - candidate.position_).lengthSquared(),
-                            candidate));
-                    }
+                        velocityObstacles_[i].apex_ + s * velocityObstacles_[i].left_side;
+                    addToCandidateListIfValid(candidate);
                 }
             }
         }
     }
 
-    int optimal = -1;
+    double min_pref_speed = pref_velocity_.length() * min_pref_speed_multiplier;
 
-    for (std::pair<float, Candidate> candidate_pair : candidates_)
-    {
-        candidate  = candidate_pair.second;
-        bool valid = true;
+    // corresponds to the velocity obstacle that is furthest away when picking the new
+    // velocity that we might collide with
+    int optimal_furthest_away_obstacle = -1;
 
+    new_velocity_ = Vector();
+    // returns the first velocity obstacle in velocityObstacles_ that the given candidate
+    // point intersects. returns -1 if it doesn't intersect any velocity obstacle
+    auto firstIntersectingVelocityObstacle = [&](Candidate &c) -> int {
         for (int j = 0; j < static_cast<int>(velocityObstacles_.size()); ++j)
         {
-            if (j != candidate.velocityObstacle1_ && j != candidate.velocityObstacle2_ &&
-                (velocityObstacles_[j].side2_)
-                        .determinant(candidate.position_ - velocityObstacles_[j].apex_) <
-                    0.0f &&
-                (velocityObstacles_[j].side1_)
-                        .determinant(candidate.position_ - velocityObstacles_[j].apex_) >
-                    0.0f)
+            if ((j != c.velocityObstacle1_ && j != c.velocityObstacle2_ &&
+                 c.intersectsVelocityObstacle(velocityObstacles_[j])) ||
+                (c.position_.length() < min_pref_speed))
             {
-                valid = false;
-
-                if (j > optimal)
-                {
-                    optimal       = j;
-                    new_velocity_ = candidate.position_;
-                }
-
-                break;
+                return j;
             }
         }
 
-        if (valid)
+        return -1;
+    };
+
+    // Choosing a candidate point has these goals:
+    // - Pick a velocity as close as possible to the preferred velocity as long as it
+    // doesn't lie in any velocity obstacle.
+    // - Failing the first condition, then choose the best preferred velocity that
+    // minimizes collisions with the closest velocity obstacle
+    // - Candidate multimap is organized by distance from preferred velocity so all we
+    // have to do is pick the first velocity that is valid
+    for (auto [dist_sq, candidate] : candidates_)
```

Comment:
nit: the structured binding here can be a const reference 
Can `dist_sq` also be updated to something like `dist_to_pref_velocity_sq`?

Reply:
updated